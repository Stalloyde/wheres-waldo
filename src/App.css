#root {
  display: flex;
  height: 100vh;
}

section {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 2em;
  gap: 5em;
  background-color: rgb(226, 218, 218);
} 


section > div {
  display: grid;
  text-align: center;
}

section > div > img {
  height: 80px;
}

.found {
  padding: 5px;
  border: 3px solid green;
  border-radius: 10px;
}

.found > img {
  height: 60px;
  opacity: 0.5;
}


main {
  display: flex;
  align-items: center;
}

#timer {
  background-color: black;
  padding: 0.2em;
  color: rgb(34, 219, 34);
  font-size: xx-large;
}

main > div {
  display: flex;
  flex-direction: column;
  align-items: center;
}

#whereswaldo {
  min-width: 800px;
  max-width: 80%;
  border: 3px solid black;
  border-radius: 10px;
  margin: 1em;
  position: relative;
} 


main > div > img:hover {
  cursor: pointer;
} 

#target {
  height:20px;
  width: 20px;
  position: absolute;
}

.wrongCharacterModal {
  position: absolute;
  background-color: rgb(226, 218, 218);
  color:rgb(187, 35, 35);
  border-radius: 5px;
  padding: 1em;
  outline: 1px solid black;
  box-shadow: 10px 10px solid black;
}