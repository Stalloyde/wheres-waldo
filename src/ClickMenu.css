ul {
  background-color: rgb(226, 218, 218);
  padding: 0.5em;
  border-radius: 5px;
  opacity: 1;
}

li {
  list-style: none;
  padding: 0.2em;
}

button {
  width: 100%;
  height: 30px;
  background-color: rgb(187, 35, 35);
  color:white;
  border-radius: 5px;
}

button:hover {
  background-color: rgb(238, 31, 31);
  cursor: pointer;
}
