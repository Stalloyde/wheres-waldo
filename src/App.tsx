import { useState } from 'react';
import usePreventZoom from './preventZoom';
import gameImage from './assets/whereswaldo.jpg';
import wizardImage from './assets/wizard.gif';
import waldoImage from './assets/waldo-standing.png';
import odlawImage from './assets/odlaw.gif';
import ClickMenu from './ClickMenu';
import target from './assets/icons8-target-32.png';
import './App.css';

function App() {
  const [click, setClick] = useState(false);
  const [mousePosition, setMousePosition] = useState({});
  const [wrongCharacterModal, setWrongCharacterModal] = useState(false);
  const [characters, setCharacters] = useState([
    {
      name: 'Wizard',
      found: false,
      src: wizardImage,
      alt: 'Wizard',
    },
    {
      name: 'Waldo',
      found: false,
      src: waldoImage,
      alt: 'Waldo',
    },
    {
      name: 'Odlaw',
      found: false,
      src: odlawImage,
      alt: 'Odlaw',
    },
  ]);

  usePreventZoom();

  function handleClick(e) {
    const xPercentage = (e.pageX / window.innerWidth) * 100;
    const yPercentage = (e.pageY / window.innerHeight) * 100;

    if (wrongCharacterModal) {
      setClick(false);
      setWrongCharacterModal(false);
    } else {
      setClick(!click);
      setWrongCharacterModal(false);
    }

    setMousePosition({
      x: xPercentage,
      y: yPercentage,
    });
  }

  function closeModal() {
    setWrongCharacterModal(!wrongCharacterModal);
  }

  const targetStyle = {
    position: 'absolute',
    top: `${mousePosition.y - 1.5}%`,
    left: `${mousePosition.x - 0.2}%`,
  };

  characters.map((char) => console.log(char));
  return (
    <>
      <section>
        <h2>Find These Characters:</h2>

        {characters.map((character) => {
          if (!character.found) {
            return (
              <div key={character.name}>
                <img src={character.src} alt={character.alt} />
                {character.name}
              </div>
            );
          } else {
            return (
              <div key={character.name} className='found'>
                <img src={character.src} alt={character.alt} />
                {character.name}
              </div>
            );
          }
        })}

        <div id='timer'>00:00:00</div>
      </section>

      <main>
        <div>
          <img
            src={gameImage}
            alt='wheres-waldo'
            id='whereswaldo'
            onClick={(e) => handleClick(e)}></img>
          {click && (
            <>
              <img style={targetStyle} src={target} alt='target' id='target' />
              <ClickMenu
                click={click}
                onClick={closeModal}
                setClick={setClick}
                mousePosition={mousePosition}
                characters={characters}
                setCharacters={setCharacters}
                wrongCharacterModal={wrongCharacterModal}
                setWrongCharacterModal={setWrongCharacterModal}
              />
            </>
          )}
        </div>
        {wrongCharacterModal && (
          <div className='wrongCharacterModal' style={targetStyle}>
            Wrong Character... Try again.
          </div>
        )}
      </main>
    </>
  );
}

export default App;
