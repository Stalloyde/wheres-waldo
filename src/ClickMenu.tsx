import './ClickMenu.css';

function ClickMenu({ setClick, mousePosition }) {
  const clickMenuStyle = {
    position: 'absolute',
    top: `${mousePosition.y + 0.5}%`,
    left: `${mousePosition.x + 0.5}%`,
  };

  function handleClick() {
    setClick(false);
  }
  return (
    <>
      <ul style={clickMenuStyle}>
        Who's this?
        <li>
          <button onClick={handleClick}>Wizard</button>
        </li>
        <li>
          <button onClick={handleClick}>Wally</button>
        </li>
        <li>
          <button onClick={handleClick}>Odlaw</button>
        </li>
      </ul>
    </>
  );
}

export default ClickMenu;
