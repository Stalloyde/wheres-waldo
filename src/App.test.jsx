import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, expect, it, test } from 'vitest';
import App from './App';
import ClickMenu from './ClickMenu';
import userEvent from '@testing-library/user-event';
import gameImage from './assets/whereswaldo.jpg';
import { act } from 'react-dom/test-utils';

describe('App', () => {
  it('renders page', () => {
    const app = render(<App />);
    expect(app).toMatchSnapshot();
  });
});

describe('Click on game image', () => {
  it('Sets mousePosition', async () => {
    const setMousePosition = vi.fn();
    const handleClickMock = vi.fn();
    const user = userEvent.setup();

    vi.spyOn(React, 'useState').mockImplementationOnce((initState) => [
      initState,
      setMousePosition,
    ]);

    handleClickMock.mockImplementation((xArg, yArg) => {
      setMousePosition({
        x: xArg + 10,
        y: yArg + 10,
      });
    });

    render(
      <img
        src={gameImage}
        alt='wheres-waldo'
        id='whereswaldo'
        onClick={() => handleClickMock(23, 42)}></img>,
    );

    const image = screen.getByRole('img');
    await act(async () => user.click(image));
    expect(setMousePosition).toHaveBeenCalledWith({ x: 33, y: 52 });
  });

  it('Sets click state', async () => {
    const setClick = vi.fn();
    const handleClickMock = vi.fn();
    const user = userEvent.setup();

    vi.spyOn(React, 'useState').mockImplementationOnce((initState) => [
      initState,
      setClick,
    ]);

    handleClickMock.mockImplementation(() => {
      setClick(true);
    });

    render(
      <img
        src={gameImage}
        alt='wheres-waldo'
        id='whereswaldo'
        onClick={() => handleClickMock()}></img>,
    );

    const image = screen.getByRole('img');
    await act(async () => user.click(image));
    expect(setClick).toHaveBeenCalledWith(true);
  });

  it('Renders ClickMenu on image click', async () => {
    const user = userEvent.setup();
    render(<App />);

    const gameImage = screen.getByAltText('wheres-waldo');
    await act(async () => user.click(gameImage));
    expect(screen.getByText(`Who's this?`)).toBeInTheDocument();
    expect(screen.getByAltText(`target`)).toBeInTheDocument();
  });
});
