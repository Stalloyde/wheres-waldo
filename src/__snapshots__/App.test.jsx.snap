// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <h2>
          Find These Characters:
        </h2>
        <div>
          <img
            alt="wizard"
            src="/src/assets/wizard.gif"
          />
          Wizard
        </div>
        <div>
          <img
            alt="wally"
            src="/src/assets/wally-standing.png"
          />
          Wally
        </div>
        <div>
          <img
            alt="odlaw"
            src="/src/assets/odlaw.gif"
          />
          Odlow
        </div>
        <div
          id="timer"
        >
          00:00:00
        </div>
      </section>
      <main>
        <div>
          <img
            alt="wheres-waldo"
            id="whereswaldo"
            src="/src/assets/whereswaldo.jpg"
          />
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <section>
      <h2>
        Find These Characters:
      </h2>
      <div>
        <img
          alt="wizard"
          src="/src/assets/wizard.gif"
        />
        Wizard
      </div>
      <div>
        <img
          alt="wally"
          src="/src/assets/wally-standing.png"
        />
        Wally
      </div>
      <div>
        <img
          alt="odlaw"
          src="/src/assets/odlaw.gif"
        />
        Odlow
      </div>
      <div
        id="timer"
      >
        00:00:00
      </div>
    </section>
    <main>
      <div>
        <img
          alt="wheres-waldo"
          id="whereswaldo"
          src="/src/assets/whereswaldo.jpg"
        />
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
